File: LeetCode/Arrays/1365-How-Many-Numbers-Are-Smaller-Than-the-Current-Number-Optimal.py
"""
Problem: 1365. How Many Numbers Are Smaller Than the Current Number
Approach: Frequency Array + Prefix Sum

Since nums[i] is between 0 and 100, use a frequency array.

1. Count the frequency of every number.
2. Build a prefix sum.
3. For each number x, count[x - 1] gives the
   number of elements smaller than x.

Time Complexity: O(n + 101) = O(n)
Space Complexity: O(101) = O(1)
"""

from typing import List


class Solution:
    def smallerNumbersThanCurrent(self, nums: List[int]) -> List[int]:
        count = [0] * 101

        # Frequency
        for num in nums:
            count[num] += 1

        # Prefix sum
        for i in range(1, 101):
            count[i] += count[i - 1]

        ans = []

        for num in nums:
            if num == 0:
                ans.append(0)
            else:
                ans.append(count[num - 1])

        return ans

File: LeetCode/Arrays/1365-How-Many-Numbers-Are-Smaller-Than-the-Current-Number-Optimal.js
/**
 * Problem: 1365. How Many Numbers Are Smaller Than the Current Number
 * Approach: Frequency Array + Prefix Sum
 *
 * nums[i] is between 0 and 100, so we can use a
 * fixed-size frequency array.
 *
 * Time Complexity: O(n + 101) = O(n)
 * Space Complexity: O(101) = O(1)
 */

/**
 * @param {number[]} nums
 * @return {number[]}
 */
var smallerNumbersThanCurrent = function(nums) {
    const count = new Array(101).fill(0);

    // Frequency
    for (const num of nums) {
        count[num]++;
    }

    // Prefix sum
    for (let i = 1; i <= 100; i++) {
        count[i] += count[i - 1];
    }

    const ans = [];

    for (const num of nums) {
        if (num === 0) {
            ans.push(0);
        } else {
            ans.push(count[num - 1]);
        }
    }

    return ans;
};